### PR TITLE
ci(playwright): auto-generate the source→spec map from spec analysis

### PR DESCRIPTION
**Pull request by chirag-madlani**
## Describe your changes

Third and last in the map-reliability sequence (after #32809 "unmapped code → full" and #32812 "hand-authored map hygiene"). Turns most of `impact-map.json` into an artifact the code shape *derives*, not a document a human has to remember to update. Hand-authored routing still wins where it exists — the generated map only fills gaps.

## Why the previous two PRs are not enough on their own

#32809 gives us a safety net (unmapped code → full suite) but the safety net is expensive — every uncovered edit pays for the whole 25-min run. #32812 filled the specific known gaps by hand, but there are ~80 specs still uncovered and the count grows every time someone adds a new spec without a matching mapping entry. This PR is the durable fix for the *drift* problem: the generator recomputes the map from the code every time, and a drift check keeps the committed file honest.

## What the generator does

`.github/scripts/generate_playwright_impact_map.py` walks each `playwright/e2e/**/*.spec.ts` and unions two signals:

**1. Import graph.** Relative `import` statements resolved transitively up to 4 hops. Catches direct type imports from `src/generated/**` and `src/enums/**` — a schema change now surfaces the exact specs that read that schema shape (e.g. `src/generated/entity/data/table.ts` → 6 specs).

**2. testId cross-reference.** Extracts every string from `page.getByTestId('X')` / `[data-testid="X"]` in the spec, matches it against `data-testid="X"` occurrences in `src/**`. A testId is a runtime contract between spec and component, so this is a much stronger signal for component/page changes than the import graph.

The generator drops testIds owned by more than 3 source files. `loader`, `close-btn`, `save-button`, and similar wrappers are used everywhere; mapping them to every use site would generate a 500-way fan-out entry on every shared-component edit.

Runs in ~2s. Output: `.github/playwright/impact-map.generated.json` (committed).

## Coverage jump

- Hand-authored map alone: **252 / 362 mappable specs (71%)**
- Generated map alone: **303 / 362 specs (78%)**
- Hand + generated together: **~340 / 362 specs (94%)**

`MetricBulkImportExportEdit.spec.ts` — the recurring merge-queue ejector — now routes from 5 sources including `pages/MetricsPage/MetricListPage/MetricListPage.tsx` and `components/BulkEditEntity/BulkEditEntity.component.tsx`. Live simulation with the generated map loaded:

```
$ echo "openmetadata-ui/…/pages/MetricsPage/MetricListPage/MetricListPage.tsx" > /tmp/changed
$ python3 .github/scripts/select_playwright_tests.py --event-name pull_request \
    --changed-files /tmp/changed \
    --impact-map .github/playwright/impact-map.json \
    --output /tmp/plan.json
$ jq '.selectors[].spec | select(test("Metric"))' /tmp/plan.json
"playwright/e2e/Features/MetricBulkImportExportEdit.spec.ts"
"playwright/e2e/Features/MetricCustomUnitFlow.spec.ts"
"playwright/e2e/Flow/Metric.spec.ts"
"playwright/e2e/Flow/MetricListSearch.spec.ts"
```

All four Metric specs selected, no fallback. Under Option A alone this same edit would go through the "unmapped code → full" path and take ~25 min instead of the ~15 min a targeted plan takes.

## Planner integration

`select_playwright_tests.py` gains an optional `--generated-impact-map` argument. When omitted, it auto-detects a sibling `impact-map.generated.json` next to the hand-authored map — so **the CI workflow does not change**. Committing the generated file is enough for it to take effect on the next PR run.

Generated mappings never carry `projects`; they default to `["auto"]` and let the shard planner route each spec to whichever project it belongs to. On collision with a hand-authored mapping the two project sets are unioned in `add_selection`, so a spec named by both still runs on every project either map claims for it.

## Drift protection

`generate_playwright_impact_map.py --check` compares the committed generated file to what the current tree would produce and exits 1 on mismatch. A new pytest,`test_committed_generated_impact_map_matches_the_generator_output`, wraps that in `test_playwright_ci_planning.py` so drift shows up locally at pytest time.

*Note*: wiring the pytest into a CI job is a separate change — the tests file exists in `.github/scripts/tests/` but is not currently invoked by any workflow. That is a pre-existing gap; I intentionally did not touch workflow YAML in this PR sequence.

## Type of change

- [x] CI change (no product code touched; drops one 6.3KB committed artifact — the generated map)

## Test coverage

99 pytests pass (94 existing + 5 new):

- `test_generated_impact_map_extends_hand_authored_routing` — a source only covered by the generated map routes to its specs; a source covered by both routes to the union.
- `test_generated_impact_map_auto_detected_beside_hand_authored` — sibling auto-detection works without an explicit CLI flag.
- `test_generator_import_graph_and_testid_signals_produce_a_stable_output` — end-to-end on a miniature UI tree, both signals resolve.
- `test_generator_drops_testids_owned_by_too_many_files` — wrapper-testId heuristic verified.
- `test_committed_generated_impact_map_matches_the_generator_output` — drift guard.

## Not part of this PR

- **Wiring the pytest into a CI job** — pre-existing gap in the repo (nothing runs `.github/scripts/tests/**` today). Should be a separate infrastructure PR.
- **Auto-regenerating the map on every push to `main`** — would require a workflow edit; the drift check catches the same class of problem on PR time.
- **URL-route cross-referencing** (`page.goto('/…')` → `pages/**`) — the two signals we already have lift coverage from 71% to 94%; a third signal would be diminishing returns for the added complexity.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by karanh37 — DISMISSED** [review]

File: .github/workflows/playwright-impact-map-drift.yml [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        if: steps.check-fork.outputs.is_fork != 'true'
+        uses: actions/checkout@v7
+        with:
+          repository: ${{ steps.pr-details.outputs.full_repo || github.repository }}
+          ref: ${{ steps.checkout-ref.outputs.ref }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          allow-unsafe-pr-checkout: true
+          persist-credentials: false
+
+      - name: Setup Python
```
[the remark this answers is not in the document]
Comment (chirag-madlani) [reply .github/workflows/playwright-impact-map-drift.yml:168]: Addressed by 5837b60. The Checkout, Setup Python, Regenerate, and Check-for-drift steps are all now gated on `steps.check-fork.outputs.is_fork != 'true'`. Fork PRs no longer check out or execute the head SHA at all — the TOCTOU pattern CodeQL flagged only fires on same-repo PRs, where the checked-out ref is our own reviewed code. Marking resolved.

File: .github/scripts/generate_playwright_impact_map.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+IMPORT_RE = re.compile(
+    r"""
+    ^\s*
+    (?:import|export)\b
+    [^'"\n]*                     # any bindings (including empty for side-effect)
+    (?:from\s+)?
+    (['"])([^'"]+)\1
+    """,
+    re.VERBOSE | re.MULTILINE,
+)
```
[the remark this answers is not in the document]
Comment (chirag-madlani) [reply .github/scripts/generate_playwright_impact_map.py:81]: Fixed in 477a5dd824. Updated `IMPORT_RE` to use lazy `[^'"]*?` under DOTALL and anchor to `(?:^|;)`, so multi-line named imports are now captured. Regenerated `impact-map.generated.json` — 553 source→spec entries after the fix (net-new sources from multi-line imports appear in the diff).

File: .github/workflows/playwright-impact-map-drift.yml [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        run: |
+          # `[skip ci]` prevents the push from re-triggering this same workflow
+          # (which would find a clean tree and pass anyway, wasting a run).
+          # PR CI's other checks still re-fire because they gate on push events,
+          # not on this workflow.
+          git commit -m "chore(playwright): auto-refresh impact-map.generated.json [skip ci]"
+          # persist-credentials: false above, so re-attach the token to origin.
+          # Bare URL would break --force-with-lease (no remote-tracking ref to
+          # lease against).
+          git remote set-url origin "https://x-access-token:${GH_TOKEN}@github.com/${{ github.repository }}.git"
```
[the remark this answers is not in the document]
Comment (chirag-madlani) [reply .github/workflows/playwright-impact-map-drift.yml:247]: Fixed in 477a5dd824. Added `(steps.pr-details.outputs.head_ref || github.event.pull_request.head.ref) != ''` to the `if:` of both Configure git and Auto-commit steps, so `workflow_dispatch` without `pr_number` becomes a clean no-op instead of pushing to `HEAD:`.

File: .github/scripts/generate_playwright_impact_map.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+RESOLVE_SUFFIXES = (".ts", ".tsx", ".d.ts")
+
+# Line-level import parser. Tolerant of `import type`, default, namespace,
+# named, and side-effect forms. Does not attempt to parse strings inside
+# comments or template literals — none of our specs use dynamic require.
+IMPORT_RE = re.compile(
+    r"""
+    (?:^|;)\s*
+    (?:import|export)\b
+    [^'"]*?                      # any bindings, may span newlines (lazy)
```
[the remark this answers is not in the document]
Comment (chirag-madlani) [reply .github/scripts/generate_playwright_impact_map.py]: Good catch — fixed in 64e92dc814. Dropped `re.DOTALL` (inert since the pattern contains no `.`).

[the remark this answers is not in the document]
Comment (chirag-madlani) [reply .github/scripts/generate_playwright_impact_map.py]: Fixed in 64e92dc814. Split the bindings alternation into two branches — side-effect `\s+` for `import 'foo'`, or `[^'";]*?\bfrom\s*` for the binding form. The `from` requirement + `;` boundary prevents `export default foo;` from lazily attributing an unrelated downstream string literal. Verified with edge-case tests including `export default foo;\nconst x = 'bar';` → no match, and multi-line `import { A,\n  B } from './x'` still captured.

**Review by ShaileshParmar11 — DISMISSED** [review]

**Review by karanh37 — APPROVED** [review]
